src/charm.py

```diff
@@ -274,24 +284,21 @@ def _prometheus_config(self) -> dict:
             ]
         }
 
-    def _loki_config(self) -> dict:
-        """Return the loki section of the config.
+    def _loki_config(self, config, event: EventBase) -> None:
+        """Modifies the loki section of the config.
 
         Returns:
-            The dict representing the config
+            None
         """
-        config = {}
+        if isinstance(event, LokiPushApiEndpointDeparted):
+            config.pop("loki", None)
 
-        if self._stored.remove_loki_config:
-            self._stored.remove_loki_config = False
-            return {}
-
-        if loki_push_api := self._loki_consumer.loki_push_api:
-            config = {
+        if isinstance(event, LokiPushApiEndpointJoined):
+            config["loki"] = {
```

Comment:
Consider recreating the entire config based on the remaining relation, instead of using isinstance and differential config update.
Also it looks like "_loki_config" could be moved into the loki lib?